src/main/java/net/consensys/orion/http/handler/send/DistributePayloadManager.java

```diff
@@ -0,0 +1,291 @@
+              future.complete(privacyGroupPayload);
+              return result;
+            }).exceptionally(e -> {
+              future.fail(new OrionException(OrionErrorCode.ENCLAVE_UNABLE_STORE_PRIVACY_GROUP));
+              return null;
+            });
+          }).exceptionally(e -> {
+            future.fail(new OrionException(OrionErrorCode.ENCLAVE_UNABLE_STORE_PRIVACY_GROUP));
+            return null;
+          });
+        }
+      }).exceptionally(e -> {
+        future.fail(new OrionException(OrionErrorCode.ENCLAVE_PRIVACY_QUERY_ERROR));
+        return null;
+      });
+    } catch (OrionException e) {
+      future.fail(e);
+    }
+
+    return future;
+  }
+
+  private Future<PrivacyGroupPayload> handlePrivacyGroupSendRequest(SendRequest sendRequest) {
+    final Future<PrivacyGroupPayload> future = Future.future();
+
+    try {
+      final String privacyGroupId = sendRequest.privacyGroupId().get();
+      privacyGroupStorage.get(privacyGroupId).thenApply((result) -> {
+        if (result.get().state().equals(PrivacyGroupPayload.State.ACTIVE)) {
+          future.complete(result.get());
+          return result;
+        } else {
+          future.fail(new OrionException(OrionErrorCode.ENCLAVE_PRIVACY_GROUP_MISSING, "privacy group not found"));
+          return result;
+        }
+      }).exceptionally(e -> {
+        future.fail(e);
+        return Optional.empty();
+      });
+    } catch (OrionException e) {
+      future.fail(e);
+    }
+
+    return future;
+  }
+
+  private Future<String> sendPayloadToParticipants(
+      final SendRequest sendRequest,
+      PublicKey fromKey,
+      final PrivacyGroupPayload privacyGroupPayload) {
+    final Future<String> future = Future.future();
+
+    try {
+      Optional<String> from = sendRequest.from();
+      final List<PublicKey> toKeys = Arrays
+          .stream(privacyGroupPayload.addresses())
+          .filter(key -> from.isEmpty() || !key.equals(from.get()))
+          .map(enclave::readKey)
+          .collect(Collectors.toList());
+
+      // toKeys = toKeys + [nodeAlwaysSendTo] --> default pub key to always send to
+      toKeys.addAll(Arrays.asList(enclave.alwaysSendTo()));
+      final Box.PublicKey[] arrToKeys = toKeys.toArray(new Box.PublicKey[0]);
+
+      // convert payload from b64 to bytes
+      final byte[] rawPayload = sendRequest.rawPayload();
+
+      // encrypting payload
+      log.debug("encrypting payload from SendRequest object");
+      final EncryptedPayload encryptedPayload =
+          enclave.encrypt(rawPayload, fromKey, arrToKeys, privacyGroupPayload.randomSeed());
+
+      final List<Box.PublicKey> keys =
+          toKeys.stream().filter(pKey -> !nodeKeys.contains(pKey)).collect(Collectors.toList());
+
+      if (keys.stream().anyMatch(pKey -> networkNodes.urlForRecipient(pKey) == null)) {
+        throw new OrionException(OrionErrorCode.NODE_MISSING_PEER_URL, "couldn't find peer URL");
+      }
+
+      // storing payload
```

Comment:
nit: this and the next comments are not overly useful - in fact, I think this one is wrong - the payload is stored once all peers have received the payload (line 269)